### PR TITLE
lib/protoparser: adds opentelemetry parser

### PR DESCRIPTION
**Pull request by f41gh7**
app/{vmagent,vminsert}: adds opentelemetry ingestion path

Adds ability to ingest data with opentelemetry protocol
protobuf and json encoding is supported
data converted into prometheus protobuf timeseries
each data type has own converter and it may produce multiple timeseries
from single datapoint (for summary and histogram).
only cumulative aggregationFamily is supported for sum(prometheus
counter) and histogram.

For https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2424

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1000,6 +1001,16 @@ Note that it could be required to flush response cache after importing historica
 
 VictoriaMetrics also may scrape Prometheus targets - see [these docs](#how-to-scrape-prometheus-exporters-such-as-node-exporter).
 
+## Sending data via opentelemetry http
+
+ VictoriaMetrics supports data ingestion with [opentelemetry protocol](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#opentelemetry-protocol-data-model) with `protobuf` and `json` encoding
```
Comment (hagen1778) [inline README.md:1006]:
```suggestion
 VictoriaMetrics supports data ingestion via [opentelemetry protocol](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#opentelemetry-protocol-data-model) with `protobuf` and `json` encoding
```

File: README.md [hunk 2]
```diff
@@ -1000,6 +1001,16 @@ Note that it could be required to flush response cache after importing historica
...
+ via `/opentemetry/api/v1/push` path. For example, following command ingest single gauge metric:
```
Comment (hagen1778) [inline README.md:1007]:
```suggestion
 via `/opentemetry/api/v1/push` path. For example, following command ingests a single gauge metric:
```

File: README.md [hunk 3]
```diff
@@ -1000,6 +1001,16 @@ Note that it could be required to flush response cache after importing historica
...
+```bash
+curl -XPOST -H 'Content-Type: application/json' localhost:8428/opentelemetry/api/v1/push -g -d '{"resourceMetrics":[{"resource":{"attributes":[{"key":"job", "value":{"stringValue":"vm"}}]}, "scopeMetrics":[{"metrics":[{"name":"my-gauge", "gauge":{"dataPoints":[{"attributes":[{"key":"label1", "value":{"stringValue":"value1"}}], "timeUnixNano":"15000000000", "asInt":"15"}]}}]}]}]}'
+```
+ By default, VictoriaMetrics decodes response with `protobuf` decoder, pass http header `Content-Type: application/json` for json encoded request.
```
Comment (hagen1778) [inline README.md:1011]:
```suggestion
 By default, VictoriaMetrics expects `protobuf`-encoded requests. For sending `json`-encoded requests set HTTP header `Content-Type: application/json`.
```

File: README.md [hunk 4]
```diff
@@ -1000,6 +1001,16 @@ Note that it could be required to flush response cache after importing historica
...
+ VictoriaMetrics accepts data with gzip compression, pass http header `Content-Encoding: gzip` for compressed data.
```
Comment (hagen1778) [inline README.md:1012]:
```suggestion
 VictoriaMetrics accepts data with gzip compression, set HTTP header `Content-Encoding: gzip` for compressed data.
```

File: app/vminsert/opentelemetryhttp/request_handler.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/VictoriaMetrics/metrics"
+)
+
+var (
+	rowsInserted  = metrics.NewCounter(`vm_rows_inserted_total{type="opentelemetry"}`)
+	rowsPerInsert = metrics.NewHistogram(`vm_rows_per_insert{type="opentelemetry"}`)
+)
+
+// InsertHandler processes opentelemetry metrics.
+func InsertHandler(req *http.Request) error {
```
Comment (hagen1778) [inline app/vminsert/opentelemetryhttp/request_handler.go:22]: does it support tenants like vmagent's handler?

**Review by hagen1778 — COMMENTED** [review]
LGTM

File: app/vminsert/opentelemetryhttp/request_handler.go (hunk 5 above)
Comment (f41gh7) [reply app/vminsert/opentelemetryhttp/request_handler.go:22]: It uses tenant token only at cluster branch

**Review by hagen1778 — APPROVED** [review]
LGTM

**Comment by valyala** [comment]
This pull request increases VictoriaMetrics binary size from 19MB to 29MB, e.g. by around 50%. This is too big price to pay for the inclusion of OpenTelemetry-compatible JSON and protobuf data ingestion protocols :( All the [existing data ingestion protocols](https://docs.victoriametrics.com/#how-to-import-time-series-data) supported by VictoriaMetrics increase binary size by a few tens of kilobytes in worst case. It would be great trying to decrease the binary size overhead for this change before merging it into VictoriaMetrics.

In the mean time it is possible to export data from [OpenTelemetry collector client](https://github.com/open-telemetry/opentelemetry-collector-contrib/) to VictoriaMetrics via one of [the supported data ingestion protocols](https://docs.victoriametrics.com/#how-to-import-time-series-data):
- [Via Prometheus pull protocol](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusexporter)
- [Via Prometheus remote_write protocol](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusremotewriteexporter)
- [Via InfluxDB protocol](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/influxdbexporter)
- [Via Graphite protocol](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/carbonexporter)
- [Via DataDog protocol](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/datadogexporter)

**Review by valyala — CHANGES_REQUESTED** [review]
See https://github.com/VictoriaMetrics/VictoriaMetrics/pull/2570#issuecomment-1221981838

**Comment by rodrigc** [comment]
@valyala might be nice to add to the VictoriaMetrics docs, some
example configs (or pointers to configs in OpenTelemetry docs) for configuring an OpenTelemetry client to push to VictoriaMetrics via the [various protocols you mentioned](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/2570#issuecomment-1221981838).  That way end-users would have a better idea how to configure OpenTelemetry to push to VictoriaMetrics.  If the existing protocols work well enough, then a native OLTP ingest path may not be necessary.

If native OLTP could be implemented without increasing the footprint requirements for VictoriaMetrics, that would be nice too, but is a longer discussion/implementation.

**Comment by syepes** [comment]
+1 Hope you can find a solution to this one, this would be a great addition.

**Comment by syepes** [comment]
@f41gh7 Is there any news on the merger of this feature? it would be great to have native support for this new protocol

**Comment by PatMis16** [comment]
Hi VictoriaMetrics Developers!
How about this feture with the OpenTelemetry protocol? Is there a planned release date for this?

**Comment by f41gh7** [comment]
Will take a look and update dependency soon.

**Comment by f41gh7** [comment]
@valyala PR was updated, but any way it increases result binary size by 5MB because of protobuf json parser dependencies. It's possible to optimize it further, but I'm not sure, that it worth it.

**Comment by valyala** [comment]
> PR was updated, but any way it increases result binary size by 5MB because of protobuf json parser dependencies. It's possible to optimize it further, but I'm not sure, that it worth it.

+5MB is better than +10MB from [the previous attempt](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/2570#issuecomment-1221981838), but is is still too much for a simple parser for metrics in OpenTelemetry protocol when comparing to binary footprints for other data ingestion protocols supported by VictoriaMetrics :(

I believe we can write our own implementation for the OpenTelemetry metrics parser, which will increase the resulting binary size by no more than 100Kb in the worst case.

**Comment by f41gh7** [comment]
@valyala  changed implementation - rewritten parser from scratch with default protobuf generator. Produced binary size almost the same as it was before adding this feature.

**Comment by syepes** [comment]
@f41gh7 Do you guys have an rough idea when you will be releasing a version with this?

**Comment by valyala** [comment]
> Do you guys have an rough idea when you will be releasing a version with this?

We had plans to merge this pull request in the upcoming v1.86.0 release, but it won't be included there because of issues with the increased binary size :( The latest version of the pull request still increases VictoriaMetrics binary size by 4MiB from 19MiB to 23MiB. I hope these issues will be resolved until the next release, so the support for OpenTelemetry protocol will be added into the next release.

Side note: the [OpenTelemetry metrics protocol specification](https://github.com/open-telemetry/opentelemetry-specification/blob/ffddc289462dfe0c2041e3ca42a7b1df805706de/specification/metrics/data-model.md) is an over-engineered nightmare comparing to other popular data ingestion formats such as [Influx line protocol](https://docs.influxdata.com/influxdb/v1.8/write_protocols/line_protocol_tutorial/), [Graphite plaintext protocol](https://graphite.readthedocs.io/en/latest/feeding-carbon.html#the-plaintext-protocol) and [Prometheus text exposition format](https://github.com/Showmax/prometheus-docs/blob/52ad0a3fd52a0c2602c3b70205578f78350eb941/content/docs/instrumenting/exposition_formats.md). This introduces non-trivial overhead to users, clients and server, which have to deal with it. For example, a simple `temperature{job="vm",label1="value1"} 15` [metric](https://docs.victoriametrics.com/keyConcepts.html#what-is-a-metric) transforms into the following monster when being transferred with OpenTelemetry metrics JSON format:
```
{
  "resourceMetrics": [
    {
      "resource": {
        "attributes": [
          {
            "key": "job",
            "value": {
              "stringValue": "vm"
            }
          }
        ]
      },
      "scopeMetrics": [
        {
          "metrics": [
            {
              "name": "temperature",
              "gauge": {
                "dataPoints": [
                  {
                    "attributes": [
                      {
                        "key": "label1",
                        "value": {
                          "stringValue": "value1"
                        }
                      }
                    ],
                    "asInt": "15"
                  }
                ]
              }
            }
          ]
        }
      ]
    }
  ]
}
```

**Comment by f41gh7** [comment]
@valyala  changed proto compiler. Without any external dependency it increases binary size for 1MB approx.

**Comment by hagen1778** [comment]
> Without any external dependency it increases binary size for 1MB approx.

not all heroes wear capes)

**Comment by rodrigc** [comment]
@hagen1778 @f41gh7 Small PR to fix the lint issues in this branch: https://github.com/VictoriaMetrics/VictoriaMetrics/pull/3943

**Comment by PatMis16** [comment]
Great work guys! Hope this is comming soon.

**Comment by syepes** [comment]
Any good news on this merger?

**Comment by zahdab** [comment]
Any update on this by any chance?

**Comment by syepes** [comment]
Closer and closer to have this merged, there will be a big party for this one 🥳

**Comment by Cho1409** [comment]
I hope to see some updates soon.

**Review by valyala — APPROVED** [review]

**Comment by valyala** [comment]
@f41gh7 , thanks for the pull request!

**Comment by hagen1778** [comment]
No way!

**Comment by valyala** [comment]
FYI, this pull request has been included in [VictoriaMetrics v1.92.0](https://docs.victoriametrics.com/CHANGELOG.html).

**Comment by sumukhs** [comment]
Hi @valyala ,

When I use `opentelemetry-sdk` python version `1.19.0` and generate a few metrics in json and try to push this to Victoria Metrics instance version v1.92.0 I see an error like this:

```
2023-08-01T22:43:01.654Z        warn    VictoriaMetrics/app/vminsert/main.go:218        remoteAddr: "172.17.0.1:45372"; requestURI: /opentelemetry/api/v1/push; cannot unpack OpenTelemetry metrics: cannot unmarshal request from 4738 bytes: proto: illegal wireType 6
```

The way I get the json data for the metrics is by dumping the state in a file like this:
```py
    reader = InMemoryMetricReader()
    provider = MeterProvider(metric_readers=[reader])
    set_meter_provider(provider)

    meter = get_meter_provider().get_meter("MIT-Metrics", "0.1.2")
    state = reader.get_metrics_data().to_json()
    logging.info(f"State:\n {state}")
    with open(output_file, "w") as f:
        json.dump(state, f)
```

**Comment by f41gh7** [comment]
@sumukhs  VictoriaMetrics supports opentelemetry only with protobuf encoding. json isn't supported atm.

**Comment by valyala** [comment]
FYI, VictoriaMetrics [v1.93.0](https://docs.victoriametrics.com/CHANGELOG.html) has the following enhancements for Opentelemetry:

1. It returns user-readable error on an attempt to push OTEL metrics with unsupported encoding. See [this pull request](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/4822).
2. It prevents from panic when pushing histograms with non-counter buckets. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4814) for details.

**Comment by valyala** [comment]
FYI, the commit dd25049858a12816f19c5f03b5c7ceb5bbe341bb switches parsing of OTEL protobuf messages from [`protoc`](https://grpc.io/docs/protoc-installation/)-generated code to [github.com/VictoriaMetrics/easyproto](https://github.com/VictoriaMetrics/easyproto). This reduces code bloat and binary bloat (VictoriaMetrics binary size has been reduced by 100KB on top of [this change](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/2570#issuecomment-1446755432)), and improves maintainability of the resulting code. This commit will be included in the next release of VictoriaMetrics.

**Pull request merged**